src/nspawn/nspawn.c

```diff
@@ -1887,14 +1887,14 @@ int userns_lchown(const char *p, uid_t uid, gid_t gid) {
 
 int userns_mkdir(const char *root, const char *path, mode_t mode, uid_t uid, gid_t gid) {
         const char *q;
-        int r;
 
         q = prefix_roota(root, path);
-        r = mkdir_errno_wrapper(q, mode);
-        if (r == -EEXIST)
-                return 0;
-        if (r < 0)
-                return r;
+        if (mkdir(q, mode) < 0) {
```

Comment:
I'd prefer to drop this commit. You are correct that mkdir_errno_wrapper() is not *necessary*. But it still is *useful* because it removes the need for error-prone manual `errno` handling. 2 out of the 4 chunks increase indentation and arguably made the code harder two read. With the other two, the code is a tiny bit shorter with `errno`, but not more readable.

Reply:
well, i thought about this when i dropped this and I think we should probably add a macro-like helper like this:

```c
static inline int NEGATIVE_ERRNO(int ret) {
    return ret < 0 ? -errno : ret;
}
```

with that in place we could do:

```c
r = NEGATIVE_ERRNO(mkdir(…));
if (r < 0)
        return r;
```

and it would work for any syscall, and not require per-syscall definitions. I think that's a much better approach than reusing the the mkdir specific wrapper that was only added to make the selinux linking mess work, and that the subsequent commit actually entirely removes in favour of mkdirat_errno_wrapper().

Opinions on such a macro? better name for it? WRAP_ERRNO()?

Reply:
Hmm, so we already have `negative_errno()` for the cases where somebody forgot to set `errno`. And I think `NEGATIVE_ERRNO` should use that too, i.e. the result from the macro should always be negative value if the function returned negative.

So far we used custom wrappers because not all functions that use `errno` return `int`… But `int` is the most popular, so it should be OK to have a wrapper just for this.

Maybe we should call this GET_ERRNO()?

Reply:
ERRNO_ERROR sounds like stuttering?

RET_ERRNO()?

ERRNO_RET()?

Reply:
#21367

Reply:
Shouldn't this use RET_NERRNO now?

Reply:
done now. ptal.